### ISSUE TITLE
[Bug] MoE aux_loss gradient never flows — MoEAuxLossAutoScaler return value discarded

### ISSUE THREAD

**Issue by LoganVegnaSHOP** [NVIDIA-NeMo/Automodel#1408]
# [Bug] MoE aux_loss gradient never flows — MoEAuxLossAutoScaler return value discarded

## Summary

There are **three compounding bugs** in the MoE load balancing loss implementation that result in the aux_loss gradient **never reaching the router weights** (and when patched to flow, having the wrong magnitude):

1. **`MoEAuxLossAutoScaler.apply()` return value is discarded** (line 372 of `layers.py`). The wrapped tensor is thrown away and the original `weights` tensor (without the autograd node) is returned from `Gate.forward()`. Since the autograd node is orphaned, `MoEAuxLossAutoScaler.backward()` is never called. The aux loss gradient is completely dead code.

2. **`aux_loss_coeff` is never applied as a scaling factor.** Even if bug 1 were fixed, the coefficient (e.g., 0.001 for Qwen3-MoE) is only used as a boolean gate (`> 0`), never multiplied with the loss.

3. **`* weights.shape[0]` (N tokens) scaling is incorrect for FSDP2 training.** The comment says "Training scales all gradients by 1/(number of tokens)" — this is true for Megatron-LM's pipeline-parallel training loop, but NeMo AutoModel's FSDP2 recipe uses a mean-reduced CE loss where gradients are NOT additionally divided by N. This `* N` factor makes the aux loss gradient ~180,000x too strong when bugs 1 and 2 are fixed.

### Bug 1: Discarded return value

We confirmed experimentally: running two training jobs — one with `main_loss_backward_scale=1.0` (default) and one with `main_loss_backward_scale=0.001` — produced **identical aux_loss curves**. Changing the backward scale has zero effect because the backward function is never invoked.

### Bug 3: `* N` scaling too aggressive

After fixing bugs 1 and 2 by monkey-patching `Gate.forward` with `weights = MoEAuxLossAutoScaler.apply(weights, self.aux_loss_coeff * aux_loss * weights.shape[0])`:

| Metric             | Without patch (aux dead) | With patch (N scaling) |
| ------------------ | ------------------------ | ---------------------- |
| grad_norm (step 0) | 38.47                    | 11,483                 |
| CE loss (step 40)  | 5.84                     | 6.12 (rising)          |
| CE loss (step 60)  | —                        | 8.57 (diverged)        |
| aux_loss (step 60) | 1.65                     | 1.22                   |

The aux loss gradient dominated training (grad_norm 300x larger), causing CE loss to diverge while aux_loss improved aggressively.

### Minimal PyTorch reproduction (Bug 1)

```python
import torch

class AuxFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, output, aux):
        ctx.save_for_backward(aux)
        return output

    @staticmethod
    def backward(ctx, grad_output):
        (aux,) = ctx.saved_tensors
        print("AuxFn.backward CALLED", float(aux))
        return grad_output, torch.ones_like(aux)

# Case 1: discard apply() return value — backward NOT called
w = torch.randn(4, requires_grad=True)
aux = (w.mean() * 0.0) + 3.14
AuxFn.apply(w, aux)          # return value discarded
loss = (w * 2).sum()          # uses original w
loss.backward()               # AuxFn.backward NOT called

# Case 2: capture apply() return value — backward CALLED
w = torch.randn(4, requires_grad=True)
aux = (w.mean() * 0.0) + 2.71
w = AuxFn.apply(w, aux)      # capture wrapped tensor
loss = (w * 2).sum()          # uses wrapped w
loss.backward()               # AuxFn.backward CALLED
```

## Affected Code

**`nemo_automodel/components/moe/layers.py`**, `Gate.forward()` lines 366-377:

```python
aux_loss = None
if self.aux_loss_coeff > 0 and self.training:
    aux_loss = self._compute_aux_loss(original_scores, expert_load, token_mask, cp_mesh)
    # Scale the aux_loss by the number of tokens.
    # Training scales all gradients by 1/(number of tokens).
    # To correct this scaling, we need to scale the aux_loss by number of tokens here.
    MoEAuxLossAutoScaler.apply(weights, aux_loss * weights.shape[0])  # <-- BUG 1: return discarded
                                                                       # <-- BUG 2: no coeff
                                                                       # <-- BUG 3: * N wrong for FSDP2

if self._track_load_balance and aux_loss is not None:
    self._last_aux_loss = aux_loss.detach()

return weights.type_as(x), indices, aux_loss  # <-- returns ORIGINAL weights, not wrapped
```

### Bug 1: Discarded return value

`MoEAuxLossAutoScaler.apply(weights, ...)` returns a new tensor with the `MoEAuxLossAutoScaler` autograd node in its backward chain. This return value is not captured. Line 377 returns the original `weights` (via `weights.type_as(x)`), which does **not** have the `MoEAuxLossAutoScaler` node. Since the wrapped tensor is never used in any downstream computation, PyTorch's autograd never calls `MoEAuxLossAutoScaler.backward()`.

### Bug 2: Missing coefficient multiplication

`self.aux_loss_coeff` is checked (`> 0`) but never multiplied with the aux loss. Combined with `MoEAuxLossAutoScaler.main_loss_backward_scale` defaulting to `1.0` (never set anywhere in the codebase), the effective coefficient would be 1.0 instead of the model's intended value (typically 0.001).

### Bug 3: Incorrect `* N` scaling for FSDP2

The `* weights.shape[0]` scaling assumes the training framework divides all gradients by N (number of tokens), as Megatron-LM does in its pipeline-parallel training loop. However, NeMo AutoModel's FSDP2 recipe (`TrainFinetuneRecipeForNextTokenPrediction`) uses `FusedLinearCrossEntropy` which returns a mean-reduced loss. Gradients are NOT additionally divided by N. The `* N` factor results in the aux loss gradient being ~N times (≈180,000x) too large relative to the CE gradient.

## Suggested Fix

The correct behavior should match HuggingFace Transformers and the original research papers, where aux loss is simply added to the total loss with a small coefficient:

```python
# HuggingFace Transformers (Qwen3MoeForCausalLM):
loss = CE_loss + router_aux_loss_coef * load_balancing_loss

# Switch Transformer paper (Fedus et al., 2021), Section 2.2:
# total_loss = CE_loss + α * N * Σ(f_i · P_i)
# where α = router_aux_loss_coef (typically 0.01 or 0.001)
# and N is the number of experts (a constant), NOT the number of tokens
```

The standard approach from the [Switch Transformer paper](https://arxiv.org/abs/2101.03961) and adopted by Mixtral, Qwen3-MoE, and DeepSeek-V2 is: `total_loss = CE + coeff * aux_loss`. The aux loss gradient flows naturally through `loss.backward()` at the correct magnitude without any special autograd tricks.

Fix line 372 to:

```python
weights = MoEAuxLossAutoScaler.apply(weights, self.aux_loss_coeff * aux_loss)
```

This:

1. **Captures the return value** so the autograd node is in the backward chain
2. **Applies the coefficient** so the gradient magnitude matches the model author's intent
3. **Removes the `* N` scaling** which is incorrect for FSDP2's mean-reduced loss

### No load balancing gradient at all

The aux_loss is computed and returned from `Gate.forward()` — it appears in logs as a reasonable value (~1.6-2.0) — but its gradient **never reaches the router weights**. Any improvement in load balance during training is purely incidental, driven by the CE loss gradient improving routing as a side effect of task learning.

### Observed during fine-tuning

We fine-tuned Qwen3-30B-A3B (128 experts, top-8, `router_aux_loss_coef=0.001`) and observed:

- **aux_loss ~2.0** (roughly 50% of experts underutilized, ideal = 1.0)
- **Broken structured outputs**: model generates malformed JSON with incorrect brace nesting in tool calls
- **Identical aux_loss curves** between `main_loss_backward_scale=1.0` and `main_loss_backward_scale=0.001`, confirming backward is never called
- **CE loss decreases normally** (6.16 → 1.2 over 140 steps), proving the model learns from CE but routing is unguided
- **When bugs 1 and 2 fixed but not 3**: grad_norm explodes 300x (38 → 11,483), CE loss diverges while aux_loss drops aggressively to 1.22

### Comparison with HuggingFace Transformers

HuggingFace explicitly adds the aux loss to the CE loss, and the gradient flows normally:

```python
# HuggingFace Qwen3MoeForCausalLM:
loss = CE_loss + router_aux_loss_coef * load_balancing_loss
```

For comparison, the same base model (Qwen3-30B-A3B) fine-tuned with TRL (which uses HuggingFace's implementation where aux loss flows correctly) produces well-formed structured outputs. The NeMo-trained model showed **50%+ degradation on real-world generative evaluations** — specifically, tool calls with deeply nested JSON arguments were malformed in over half of conversations, with the model losing track of brace depth and leaking fields to incorrect nesting levels. This level of degradation on a production fine-tuning task is directly attributable to the absence of load balancing gradient, which prevents expert specialization during training.

### Results after applying workaround

With all three bugs fixed (captured return value, coefficient applied, `* N` removed), same model, same data, same hyperparameters:

| Metric                   | Buggy (aux dead) | Fixed (coeff=0.001)        |
| ------------------------ | ---------------- | -------------------------- |
| aux_loss at convergence  | ~1.6             | **~1.2-1.4**               |
| Estimated active experts | ~80 / 128 (62%)  | **~91-107 / 128 (71-83%)** |
| grad_norm                | ~38 (CE only)    | ~27-30 (CE + gentle aux)   |
| CE loss convergence      | Normal           | Normal (slightly faster)   |

The fix recovers an estimated ~10-30 additional active experts with zero impact on CE loss convergence or training stability.

Aux Loss Curve comparing the 3 bugs:
<img width="1916" height="1064" alt="Image" src="https://github.com/user-attachments/assets/fd330c9f-42b9-4776-aaa6-c6f9f324ff33" />

CE Loss Curve comparing the 3 bugs:
<img width="1919" height="1070" alt="Image" src="https://github.com/user-attachments/assets/1f98da23-6d3f-477b-aa7c-5d7db836a10e" />

## Affected Models

Any MoE model trained with NeMo AutoModel that uses the `Gate` class with `aux_loss_coeff > 0`, including all Qwen3-MoE, Qwen3.5-MoE, and other models using this code path.

## Workaround

Monkey-patch `Gate.forward` to capture the return value, apply the coefficient, and remove the `* N` scaling:

```python
from nemo_automodel.components.moe.layers import Gate
from nemo_automodel.components.moe.megatron.moe_utils import MoEAuxLossAutoScaler

_original_forward = Gate.forward

def _patched_forward(self, x, token_mask, cp_mesh=None):
    weights, indices, aux_loss = _original_forward(self, x, token_mask, cp_mesh)
    if aux_loss is not None and self.aux_loss_coeff > 0 and self.training:
        weights = MoEAuxLossAutoScaler.apply(
            weights, self.aux_loss_coeff * aux_loss
        )
    return weights, indices, aux_loss

Gate.forward = _patched_forward
```

Note: This results in `MoEAuxLossAutoScaler.apply()` being called twice per forward (once in the original dead code, once in the patch). The original call is harmless (orphaned node, never backward-ed). Only the patched call has effect.

## Environment

- NeMo AutoModel: 0.3.0 (r0.3.0)
- Container: `nvcr.io/nvidia/nemo-automodel:25.11.00`
- Model: Qwen3-30B-A3B with EP=4, CP=8, FSDP2

**Comment by qiaochuz-nv** [comment]
## QA RCCA Analysis

**Fix PR:** #1412
**Root Cause:** Three compounding bugs in MoE auxiliary loss: (1) MoEAuxLossAutoScaler return value discarded, (2) gradient scaling wrong, (3) aux_loss gradient never flowed back. Net effect: load balancing across experts was effectively disabled.

**Trigger Config:** Any MoE model with aux_loss enabled (all MoE training)

**Coverage Assessment:**

| Dimension | Bug Config | Nearest Test | Match |
|-----------|-----------|--------------|-------|
| MoE training | Required | 14 MoE tests | Yes |
| aux_loss gradient | Required | None assert gradient flow | No |
| Load balancing | Required | No metric assertion | No |

**Status:** PARTIAL — 14 MoE tests exist and exercise the code path, but none validate that aux_loss gradient actually flows. This was a silent correctness bug — training completed but expert utilization was unbalanced.

**Conclusion:** Silent correctness gap. Existing MoE tests run the code path but don't assert on aux_loss gradient flow or expert utilization metrics. Functional smoke tests inherently can't catch this class of bug.
